### PR TITLE
Added Control of link up events by interruptions

### PR DESCRIPTION
**Pull request by Ktmi**
This PR adds in an event listener for the new interruptions API to allow kytos-ng/topology to produce the appropriate link up and down events, as well as the topology notify update event. Indirectly, this should resolve the following issues

 - #62
   + Maintenance must now specifies all affected devices, and only temporarily transforms them through status_funcs
 - #63
   + Processing of interruptions automatically calls `notify_topology_update()`
 - #64
   + Interruptions work based on ids and not objects, so no non-existent links generated.
 
 # Additional Notes
While I have added visibility to `status` and `status_reason` into the UI for Links and Switches, this feature is missing from Interfaces. Additionally, changes to the status/status_reason are not registered unless the page is reloaded. This applies to changes caused by enabling/disabling a device, or causing an interruption at the device.

### CODE REVIEW

File: main.py
Comment (viniarck) [inline main.py:1195]: @Ktmi, could you add a conditional safe guard checking if `link.status == EntityStatus.UP and not link.status_reason` before calling `self.notify_link_status_change(link, interrupt_type)`? Same [approach as we've been doing so far](https://github.com/kytos-ng/topology/blob/master/main.py#L595-L596), it turns out that `notify_link_status_change` without a safe guard checking for the status upfront it could send an unnecessary `kytos/topology.link_up` event which could have network converge implications on `mef_eline` (although it's going to try to discard redundant events) . We've been trying to maintain this consistently on topology to avoid sending redundant `kytos/topology.*` to simplify for subscribers and also to facilitate for us to follow the events when potentially debugging.

File: main.py
Comment (viniarck) [inline main.py:1174]: Same comment as below (check out that one first where I elaborated on the reason why). Could you add a safe guard `link.status == EntityStatus.DOWN and len(link.status_reason) == 1`? Just so it only sends a `kytos/topology.link_down` for the first interruption, since `kytos/topology.link_down` is a higher level event.

**Review by viniarck — COMMENTED** [review]
@Ktmi, overall looks good, it's nice that you also took the opportunity to augment some of the UI components too. I asked a few minor changes though. Also:

1 - Let me confirm the understanding regarding https://github.com/kytos-ng/topology/issues/62, looks like this one here will only be covered when the `on_*_maintenance_*` methods are nuked, which we will only be able to replace when the rest of the `topology.interruption.start|end` producers (`maintenance` in this case) is in place, right? Do you want to create a new task to make this explicit or will you reuse the same one? 

2 - Once the comments are addressed, I'd be OK merging this one here since  `topology.interruption.start|end` producer hasn't landed yet, so it should be safe, but before also merging, let me know which local test you minimally explored/ran.

File: main.py
Comment (Ktmi) [reply main.py:1174]: The problem is that we might potentially miss a link_down event, if we put guards around it like this. We could have a situation where multiple interruptions occur on a single device at the same time, resulting in multiple values in `status_reason` before we check its current state.

What we could do is have a set of links which we have sent out state messages for, and use that to ensure consistency.

Comment (viniarck) [reply main.py:1174]: Indeed, @Ktmi. Right. Yes, Keeping track of the last status for each link and only send the event if it's different. 

`self.notify_link_status_change` is being called in multiple places currently, maybe we could evolve it into `self.try_to_notify_link_status_change` where it'd have this statefullness with a dict that you can introduce, we'd be increasing it's responsibility at the expense of keeping the logic in a single place, seems great to me. We'll need lock protection there too to be safe.

I also think that we can get away without having to store this extra last state in the database and only keep it at runtime since during start up when a switch and its interfaces are loaded [interfaces will always be deactivated](https://github.com/kytos-ng/topology/blob/master/main.py#L181) since it always driven by the protocol since it's not safe to assume that after a restart it'd be active again. So, this dict could assume that the initial status is `DOWN` during a startup. With that said, this would need a bit of exploratory test since link notification is used in many places, maybe we can deliver this PR with potential  redudant `kytos/topology.link_up|link_down` and only address this part later on `2023.1.1` just so we have proper time to explore this part in a new PR? Wdyt? 

With this approach there's also opportunity to remove `last_status_is_active` but that one let's leave just for `2023.2` to avoid increasing the scope: https://github.com/kytos-ng/topology/issues/122 https://github.com/kytos-ng/topology/issues/122

File: main.py
Comment (Ktmi) [reply main.py]: Should be fixed in latest commit.

File: ui/k-info-panel/link_info.kytos
Comment (Ktmi) [reply ui/k-info-panel/link_info.kytos:58]: Added issue kytos-ng/ui#48.

File: main.py
Comment (viniarck) [inline main.py:1038]: We're still implementing `status_reason` gradually, liveness still hasn't implemented it yet. I'll recommend that you rely on `link.status` at this time, if you could refactor here. 

Also, we'll need a thread lock since this is accessed from multiple threads and self.link_down is being read and written to avoid corner cases. 

Another case to simulate would be an interfaces that goes into maintenance, controller restart, and then see if it'll potentially send another link down, if it does this will need to change and self.link_down would need to be initialized with all existing links once topology is loaded or to simplify even more maybe you could only keep track of `link_up` instead? That way if it's not there it's not UP, that seems to facilitate.

Comment (viniarck) [reply main.py:1038]: Looking forward to the local tests and e2e tests outcome later on.

**Review by viniarck — CHANGES_REQUESTED** [review]
Asked changes on https://github.com/kytos-ng/topology/pull/136/files#r1231527760

File: main.py
Comment (viniarck) [inline main.py:1040]: Neat, @Ktmi. My only final concern here is multi-threaded access. To ensure that only a single thread will be able to enter in these conditionals at a time. On `topology` we've also been using `_links_notify_lock[link.id]` to notify only once in certain places, but trying to reaquire would lead to issues, so either it could be reused/restructure it or maybe you can introduce a new lock "companion" to protect `self.link_up` or something else equivalent.

File: main.py
Comment (Ktmi) [reply main.py:147]: For handling link up, we can use execute to periodically poll for when to end the cool down caused by the link up timer. We can then generate interruptions relating to this cool down; entering links into cool down when interrupted, starting the cool down timer when no other interruptions are applied, and ending the cool down with the execute method. 

Additionally, for all methods that affect status within topology we should be having those generate interruptions.

Comment (viniarck) [reply main.py:147]: @Ktmi, if you could capture another issue for link up timer to also support status_reason similar to https://github.com/kytos-ng/topology/issues/131, and then we can continue the discussion there. Yes, we'll need this cool down following a similar pattern aligned with the timer. Notice though that currently, the cooldown for link up is only for link enabling and when handling link activations the cooldown doesn't restart for other interruptions type like liveness or mw once they go up. This is a relatively old requirement, it might evolve, but currently that's how we have it. It was introduced only to amortize excess of link-flap up notifications.

Comment (viniarck) [reply main.py:147]: Getting back to the point that I raised though. `self.link_up.add(link.id)` doesn't look coherent here, could you remove it? Initially, all `GenericEntity` will be `active: False` when loaded. 

Other than that I'm also waiting for your feedback on how you locally explored it, and if e2e tests are passing, let me know. After that I can potentially explore it too.

File: main.py
Comment (viniarck) [inline main.py:1175]: @Ktmi considering the point that @italovalcy has also brought to the attention on yesterday's meeting, maybe on both `handle_interruption_start` and `handle_interruption_end` it would good points to centralize `log.info` entries, after all, we're also using `log.info` for port status changes, so here it'd reasonable too. Another benefit of centralizing here on `topology` is that it could simplify certain logs for interruption sources. Wdyt?

**Review by viniarck — APPROVED** [review]
@Ktmi, I'll leave this PR pre-approved, looks good. 

I've also taken the opportunity to explore it locally, I ran these tests:

- I confirmed that with liveness it wouldn't redundantly notify up again when it gets enabled. 
- I also use the same link-flap script that I used on my previous topology PR, no anomalies observed, the final states were as expected. 

Once e2e tests are passing and if your local exploratory-ish tests are passing as well we can merge this. Btw, I raised a new point regarding introducing a `log.info` to get your thoughts on that, but that can also get delivered in a subsequent PR if we want to add it.

**Comment by viniarck** [comment]
> I haven't run e2e tests yet as I kept realizing several potential issues with consistency with link activation/deactivation, that I ended up rewriting a large portion of topology, and am starting to creep into other NApps. 

@Ktmi, I ran [e2e on this PR that's on top of your branch](https://github.com/kytos-ng/topology/pull/146), it's passing. It's great that you're analyzing and putting thoughts into it, but as usual, let's keep moving and addresses tasks based on their priorities and leave for `2023.2` what we can afford to (liveness for instance). If you encounter new issues or unplanned issues - or have ideas where things can be improved - please create a task and or suggestion just so it can be aligned and prioritized, by all means, feel free to use GitHub as you need to map these tasks/efforts.

>A few notes on the matter:
> 
> * of_lldp liveness will be rewritten to manage liveness status without using metadata and will instead produce interruptions

Right. But, let's leave [this one](https://github.com/kytos-ng/topology/issues/131) for `2023.2` and also compare with liveness BFD, maybe liveness BFD will still needs to be prioritized first.

> * of_core should probably be in full control of link activation, rather than topology, as it has proper visibility into the space

Topology `self.links` certainly could be moved to Kytos core since it's a shared core resource (analogous to `controller.switches` and `switch.interfaces` dicts), and then `of_core` could indeed also take the opportunity to activate. But still needs a discussion to see if it's really worth it and when we could afford to. Other than that, race conditions that were impacting have been fixed. Historically, `of_core` was also designed to be as lightweight as possible, but I agree that activating links there since it's performing certain locks could simplify and make it easier to maintain. 

> * Setting status to != UP for switches and interfaces shouldn't deactivate links, instead links should inherit the status and status reason of its parent objects

Indeed `active` is independent and `status` is a factor of `self._active` and `self._enabled`, and other logical attrs that we're registering. Regarding `status` of composed objects, yes, ideally, but again, let's plan incrementally here, let's avoid significant rewrites on parts where we can still move incrementally, and then plan accordingly on which parts we want to keep improving in the next planning, OK?

File: main.py
Comment (viniarck) [reply main.py:1175]: @Ktmi let me know about this one here, if you agree and also if it'll be part of this PR or a subsequent one.

**Comment by Ktmi** [comment]
> > I haven't run e2e tests yet as I kept realizing several potential issues with consistency with link activation/deactivation, that I ended up rewriting a large portion of topology, and am starting to creep into other NApps.
> 
> @Ktmi, I ran [e2e on this PR that's on top of your branch](https://github.com/kytos-ng/topology/pull/146), it's passing. It's great that you're analyzing and putting thoughts into it, but as usual, let's keep moving and addresses tasks based on their priorities and leave for `2023.2` what we can afford to (liveness for instance), and only keep on `2023.1.1` the rest of maintenance tasks that indeed are part of a minimal v1. If you encounter new issues or unplanned issues - or have ideas where things can be improved - please create a task and or suggestion just so it can be aligned and prioritized, by all means, feel free to use GitHub as you need to map these tasks/efforts.
> 
> > A few notes on the matter:
> > 
> > * of_lldp liveness will be rewritten to manage liveness status without using metadata and will instead produce interruptions
> 
> Right. But, let's leave [this one](https://github.com/kytos-ng/topology/issues/131) for `2023.2` and also compare with liveness BFD, maybe liveness BFD will still needs to be prioritized first.
> 
> > * of_core should probably be in full control of link activation, rather than topology, as it has proper visibility into the space
> 
> Topology `self.links` certainly could be moved to Kytos core since it's a shared core resource (analogous to `controller.switches` and `switch.interfaces` dicts), and then `of_core` could indeed also take the opportunity to activate. But still needs a discussion to see if it's really worth it and when we could afford to. Other than that, race conditions that were impacting have been fixed. Historically, `of_core` was also designed to be as lightweight as possible, but I agree that activating links there since it's performing certain locks could simplify and make it easier to maintain.
> 
> > * Setting status to != UP for switches and interfaces shouldn't deactivate links, instead links should inherit the status and status reason of its parent objects
> 
> Indeed `active` is independent and `status` is a factor of `self._active` and `self._enabled`, and other logical attrs that we're registering. Regarding `status` of composed objects, yes, ideally, but again, let's plan incrementally here, let's avoid significant rewrites on parts where we can still move incrementally, and then plan accordingly on which parts we want to keep improving in the next planning, OK?

Yeah, I guess we can put those additional changes for a later release, and just push out what we have already. For now, lets get this merged.

**Comment by Ktmi** [comment]
Getting failing e2e tests with this and kytos-ng/maintenance#78. [Here](https://gist.github.com/Ktmi/a2e2a7df47c1cfa292cfe0dd6a357999) are the test results. It seems that the evcs being properly notified.

**Comment by viniarck** [comment]
> ~Getting failing e2e tests with this and [kytos-ng/maintenance#78](https://github.com/kytos-ng/maintenance/pull/78). [Here](https://gist.github.com/Ktmi/a2e2a7df47c1cfa292cfe0dd6a357999) are the test results. It seems that the evcs being properly notified.~
> 
> I'm an idiot, the issue was on maintenance side. The maintenance events were using an incorrect endpoint.

Good morning, let's merge this one then, and then once MW PRs is also ready we can make a decision if it'll be on `2023.1` or `2023.1.1` (with the rest of MW related tasks that you mentioned like preview and so on). Thanks, David.

**Pull request merged**
